packages/react-dom/src/client/ReactDOMComponent.js

```diff
@@ -417,11 +418,19 @@ export function createElement(
         type,
       );
     }
-
     if (type === 'script') {
       // Create the script via .innerHTML so its "parser-inserted" flag is
       // set to true and it does not execute
       const div = ownerDocument.createElement('div');
+      if (__DEV__) {
+        warningWithoutStack(
+          false,
+          'Encountered script tag while rendering React component. ' +
+            'Scripts inside React components are parser inserted into document ' +
+            'and they are never executed. Furthemore rendering script nodes ' +
+            'inside components breaks when using Trusted Types.',
+        );
+      }
```

Comment:
Thanks for asking the community about this. 

Unfortunately, there is no workaround for Trusted Types (we want the content of the element to be TrustedScript unconditionally). Also in this particular case, the problem is the assignment to innerHTML of the div element.

But there is a workaround for not using the script tag. From the community responses I see, that the only legitimate use case is to have something in the DOM (which is not executed, but [some widgets may need the data](https://mobile.twitter.com/andytuba/status/1159984414681645057)). For this use case users could use [template element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) instead (it is supported everywhere apart from IE).

Warning only in cases when the script is should execute is not that easy. For example, there is also a `language` attribute which affects whether the script should be executed or not. You can read more about this in the [HTML scripting spec](https://html.spec.whatwg.org/multipage/scripting.html).

However, this warning is not only for Trusted Types and I think it might make sense to do it in a separate PR, too. WDYT?

The last thing I want to point out, that this is scripts don't execute only on client side. However on SSR they work (as they are just rendered to html string). I see some people use this ([ref1](https://mobile.twitter.com/jaquers/status/1160017562442305536) and [ref2](https://mobile.twitter.com/naholyr/status/1160101438468628480)). I see the use case for this, however this is fundamentally different from client side where the script is not executed. 

P.S.: I see that a lot of people were surprised by the fact that React doesn't execute scripts (me included) and this suggests that the use case is rare.

Reply:
If we put the warning behind a feature flag (together with other code), that buys us time to consider the implications without making a final call now.

Reply:
I think it's ready :) Can you have a look if I didn't miss something?